File: step1_start_project/src/main/java/migration/simple/repository/UserRepository.java
package migration.simple.repository;

import migration.simple.config.DBConfiguration;
import migration.simple.types.User;
import org.springframework.stereotype.Repository;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Optional;

@Repository
public class UserRepository {
    private DBConfiguration.DbConfig dbConfig;

    public UserRepository(DBConfiguration.DbConfig dbConfig) {
        this.dbConfig = dbConfig;
    }

    private Long index = 3L;

    private List<User> users = Arrays.asList(
            new User(1L, "Oleg", "BigMan", 21),
            new User(2L, "Lesia", "Listova", 25),
            new User(3L, "Bin", "Bigbanovich", 30)
    );

    public List<User> findAllUsers() {
        return new ArrayList<>(users);
    }

    public synchronized boolean addUser(User newUser) {
        return users.add(newUser.copy(nextIndex()));
    }

    public Optional<User> findUser(Long id) {
        return users.stream()
                .filter(user -> user.getId().equals(id))
                .findFirst();
    }

    public synchronized boolean deleteUser(Long id) {
        Optional<User> findUser = users.stream()
                .filter(user -> user.getId().equals(id))
                .findFirst();

        Boolean status = false;
        if (findUser.isPresent()) {
            users.remove(findUser.get());
            status = true;
        }

        return status;
    }

    private Long nextIndex() {
        return index++;
    }
}

File: step3_migration_to_kotlin/src/main/kotlin/migration/simple/repository/UserRepository.kt
package migration.simple.repository

import migration.simple.config.DBConfiguration
import migration.simple.types.User
import org.springframework.stereotype.Repository

@Repository
open class UserRepository(dbConfig: DBConfiguration.DbConfig) {

    private var index: Long = 3L

    private val users = mutableListOf(
            User(1L, "Oleg", "BigMan", 21),
            User(2L, "Lesia", "Listova", 25),
            User(3L, "Bin", "Bigbanovich", 30)
    )

    init {
        println(dbConfig)
    }

    open fun findAllUsers(): List<User> {
        return users.toList()
    }

    @Synchronized
    open fun addUser(newUser: User): Long? {
        val newIndex = nextIndex()
        val addStatus = users.add(newUser.copy(id = newIndex))

        return if (addStatus) {
            newIndex
        } else {
            null
        }
    }

    open fun findUser(id: Long): User? {
        return users.firstOrNull { user -> user.id == id }
    }

    @Synchronized
    open fun deleteUser(id: Long?): Boolean {
        val findUser = users.find { user -> user.id == id }

        val addStatus: Boolean = if (findUser != null) {
            users.remove(findUser)
            true
        } else {
            false
        }

        return addStatus
    }

    private fun nextIndex(): Long {
        return index++
    }
}
